compiler/src/iree/compiler/Modules/IO/Stream/IR/IOStreamOps.td

```diff
@@ -0,0 +1,370 @@
+  let description = [{
+    Reads and returns a single byte from the stream and a booleaning indicating
+    whether the end-of-stream was reached. If the end-of-stream was reached the
+    byte returned will have undefined contents.
+  }];
+
+  let arguments = (ins
+    IOStream_HandleType:$handle
+  );
+  let results = (outs
+    I8:$result,
+    I1:$eos
+  );
+
+  let assemblyFormat = [{
+    `(` operands `)` attr-dict `:`
+    `(` type(operands) `)` `->` `(` type(results) `)`
+  }];
+
+  let extraClassDeclaration = [{
+    void getAsmResultNames(function_ref<void(Value, StringRef)> setNameFn) {
+      setNameFn(getResult(), "byte");
+      setNameFn(getEos(), "is_eos");
+    }
+  }];
+}
+
+def IOStream_ReadBytesOp : IOStream_Op<"read.bytes", [
+  OpAsmOpInterface,
+  SameVariadicOperandSize,
+]> {
+  let summary = [{reads a span of bytes into the target buffer}];
+  let description = [{
+    Reads up to the specified length into the buffer at the given offset.
+    Returns the total number of bytes read which may be less than the requested
+    length if the stream reaches its end prior to completing the read. When an
+    end-of-stream is hit all bytes until then are populated and the remaining
+    range between the read bytes and requested length will be unmodified.
+  }];
+
+  let arguments = (ins
+    IOStream_HandleType:$handle,
+    Util_BufferType:$buffer,
+    Optional<Util_Offset>:$offset,
+    Optional<Util_Offset>:$length
+  );
+  let results = (outs
+    Util_Offset:$result
+  );
+
+  let assemblyFormat = [{
+    `(` operands `)` attr-dict `:`
+    `(` type(operands) `)` `->` type($result)
+  }];
+
+  let extraClassDeclaration = [{
+    void getAsmResultNames(function_ref<void(Value, StringRef)> setNameFn) {
+      setNameFn(getResult(), "read_length");
+    }
+  }];
+}
+
+def IOStream_ReadDelimiterOp : IOStream_Op<"read.delimiter", [
+  OpAsmOpInterface,
+]> {
+  let summary = [{reads until the given delimiter byte is reached}];
+  let description = [{
+    Returns a buffer containing all bytes between the current stream offset and
+    the first occurrence of the given delimiter byte (exclusive) or the
+    end-of-stream. Returns an empty buffer if the next byte in the stream is
+    the delimiter. Returns null if the stream is already at end-of-stream and
+    no more bytes are available. The result does not contain the delimiter.
+  }];
+
+  let arguments = (ins
+    IOStream_HandleType:$handle,
+    I8:$delimiter
+  );
+  let results = (outs
+    Util_BufferType:$result
+  );
+
+  let assemblyFormat = [{
+    `(` operands `)` attr-dict `:`
+    `(` type(operands) `)` `->` type($result)
+  }];
+
+  let extraClassDeclaration = [{
+    void getAsmResultNames(function_ref<void(Value, StringRef)> setNameFn) {
+      setNameFn(getResult(), "read_bytes");
+    }
+  }];
+}
+
+def IOStream_ReadLineOp : IOStream_Op<"read.line", [
+  OpAsmOpInterface,
+]> {
+  let summary = [{reads until the next newline character is reached}];
```

Comment:
Is this equivalent to read.delimiter with newline as the delimiter? Would "pseudo-op" be an accurate description here, or is that term only really used for the VM?